### PR TITLE
Feat/email login migration

### PR DESCRIPTION
**Pull request by alexanderleegs**
# Problem
When moving users over from github login to email login, we need to cross reference their Github username with their internal user id, in order to add them as a site member. To aid this process, this script is run to migrate existing users from the github login flow to the email login flow. We separate out the database population step from the revoking of github access so we can switch the login flow only when the agency is due to be onboarded onto the new flow.

This script connects directly to our prod database using the same setup as isomercms-backend - there's some duplication of code involved, but we opted to add this as a script here instead of an endpoint on our backend as this is meant to be a one-off migration, rather than something we permanently expose on our backend.

### CODE REVIEW

File: src/emailLogin/removeGithubAccess.js
Comment (seaerchin) [inline src/emailLogin/removeGithubAccess.js]: suggest tracking the `site` also for this to be useful

File: src/emailLogin/runMigration.js
Comment (seaerchin) [inline src/emailLogin/runMigration.js:8]: suggest removing: 

```ts
'rc-davis', 'lamkeewei', 'pallani', 'LoneRifle', 

'prestonlimlianjie', 'lisatjide', 'kwajiehao', 

'gweiying', 'audreytcy', 'yanjunquek', 

'chloe-opengovsg', 'lennardl', 'oliverli', 'taufiq' 
```
as they don't seem to be part of ogp/isomer/have a valid use case

File: src/emailLogin/runMigration.js
Comment (seaerchin) [inline src/emailLogin/runMigration.js]: just destructure directly: 

```ts 
.map(({ login }) => login)
```

File: src/emailLogin/runMigration.js
Comment (seaerchin) [inline src/emailLogin/runMigration.js]: why do we only take the first value?

File: src/emailLogin/runMigration.js
Comment (seaerchin) [inline src/emailLogin/runMigration.js]: this isn't path agnostic sadly enough but idk if it's worth the effort to change it

File: src/emailLogin/runMigration.js
Comment (seaerchin) [inline src/emailLogin/runMigration.js]: this comment is confusing cos people might think it belongs to the `catch` block below

File: src/emailLogin/runMigration.js
Comment (seaerchin) [inline src/emailLogin/runMigration.js]: wouldn't this be caught by the overarching catch below?

File: src/emailLogin/runMigration.js
Comment (seaerchin) [inline src/emailLogin/runMigration.js]: suggest using `console.error` and logging the site

File: src/emailLogin/removeGithubAccess.js
Comment (seaerchin) [inline src/emailLogin/removeGithubAccess.js]: clarifying question - this does set the permission to `pull` but do the original permissions still remain for the already added members of the site?

File: src/emailLogin/runMigration.js
Comment (seaerchin) [inline src/emailLogin/runMigration.js:1]: suggest renaming this to `generateMigrations` instead to avoid confusion

File: src/emailLogin/removeGithubAccess.js
Comment (alexanderleegs) [reply src/emailLogin/removeGithubAccess.js]: Makes sense, https://github.com/isomerpages/isomer-conversion-scripts/pull/56/commits/c8915b54561821d381a7e25579dac19790f12962

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js:8]: These were historical users that might possibly still be in agency teams, particularly those who were around when isomer was first conceived - I think it doesn't matter too much either way, but it might be neater if we remove these members explicitly, what are your thoughts?

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js]: Done in https://github.com/isomerpages/isomer-conversion-scripts/pull/56/commits/8040c4678a8cf48dc0edf27fc2f8dba7bdfebf4c

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js]: RepoData.name is be the github name of a repo, which is unique - so there should only be a single item returned from the query

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js]: This is mostly for internal data keeping, do you think we should change it?

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js]: Removed in https://github.com/isomerpages/isomer-conversion-scripts/pull/56/commits/8040c4678a8cf48dc0edf27fc2f8dba7bdfebf4c!

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js]: I put this in a separate catch block because this only wraps the log creation - it would still be successful without it, what do you think?

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js]: Makes sense, https://github.com/isomerpages/isomer-conversion-scripts/pull/56/commits/8040c4678a8cf48dc0edf27fc2f8dba7bdfebf4c

File: src/emailLogin/removeGithubAccess.js
Comment (alexanderleegs) [reply src/emailLogin/removeGithubAccess.js]: This sets the access permissions of the team to `read`, and the members have access to the site only under this team, so their permissions will also be affected

File: src/emailLogin/README.md
Comment (kishore03109) [inline src/emailLogin/README.md:20]: checking in here, this isnt intended right, since this was the previous migration instruction? 
I am ok with having mutliple README, but can I get your help to just link this from the main project root's readme?

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js:1]: We actually run the migrations though, just that we additionally log the insert query in case we need to revert for any reason, do you think the name is clear enough?

File: src/emailLogin/runMigration.js
Comment (kishore03109) [inline src/emailLogin/runMigration.js:43]: should we log/throw some error here? This would be unexpected error that we would have to reconcile right?

File: src/emailLogin/removeGithubAccess.js
Comment (kishore03109) [reply src/emailLogin/removeGithubAccess.js]: Hmm, I am confused by this. Shouldnt we remove the team entirely? ie, the only team with any access to the repo should be core?

File: src/emailLogin/README.md
Comment (alexanderleegs) [reply src/emailLogin/README.md:20]: This is still needed! The first part is setup and where to retrieve credentials, this part is actually running it. I've also updated the main README in https://github.com/isomerpages/isomer-conversion-scripts/pull/56/commits/7dce77b4dcb7a4624132a46da243cdcd929b4f8e

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js:43]: Yeah that's a good point, I've added an error message to double check the team name in https://github.com/isomerpages/isomer-conversion-scripts/pull/56/commits/877482372a73406d5f2d5605b32f7a5ec558744e

File: src/emailLogin/removeGithubAccess.js
Comment (alexanderleegs) [reply src/emailLogin/removeGithubAccess.js]: I think that's also possible! I opted to change their access to readonly so that we can still check on Github who the members of the team originally were - do you think we should remove the team entirely?

File: src/emailLogin/runMigration.js
Comment (seaerchin) [reply src/emailLogin/runMigration.js:8]: if so, could we file an issue to write a separate script for removing? 

my understanding is that their permissions as-is would be greater than their permissions after the script ran. is my understanding correct?

File: src/emailLogin/runMigration.js
Comment (seaerchin) [reply src/emailLogin/runMigration.js]: this seems like an important assumption - we should write this down in code + throw an error if there are 2 identical github names rather than have it silently proceed on

File: src/emailLogin/runMigration.js
Comment (seaerchin) [reply src/emailLogin/runMigration.js]: yea then idts HAHAA just pointing it out

File: src/emailLogin/runMigration.js
Comment (seaerchin) [reply src/emailLogin/runMigration.js:1]: my bad; i think `runMigration` is ok assuming that we don't ahve other email login related migrations

File: src/emailLogin/runMigration.js
Comment (seaerchin) [reply src/emailLogin/runMigration.js]: this is a case where it works but it runs counter to clarity; i think what would be good here is: 

1. determine if email part is important enough to fail (in here, we don't so clearly not AHHA)
2. shift some stuff out (thinking the second `try` block) into another function so that it's clearer what the block does

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js:8]: It depends on how the agency team was set up - from past experience, I've seen some repos where the only team that has write access is the agency's team, instead of the more general isomeradmin + agency github teams that we do for repos now - so in these cases, the isomer members are present directly on the agency teams, and permissions will be downgraded accordingly after the script is run

I was thinking that it's not necessary to remove these users from the github teams since all access will be cut off soon - what I wanted to ensure was that they don't accidentally get added to the site member list we maintain in our own internal db

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js]: yeah that's a good point, added an error and early termination in https://github.com/isomerpages/isomer-conversion-scripts/pull/56/commits/9843d960a98dc97b6891f75231a3a57492e7a23f

File: src/emailLogin/removeGithubAccess.js
Comment (kishore03109) [reply src/emailLogin/removeGithubAccess.js]: I have a preference for removing as it is a cleaner solution! Also am worried about the edge case of: 
1. Agency has user X and user Y in their team. 
2. User X leaves the agency.
3. Agency requests to privatise their repos, repo is becomes private.
4. User X's github account still has read access to the repo. 

In essence, I think source of truth should always come from our db.
I think for logs, we can dump it to a file? and just put somewhere in our isomer folder if we need to parse it for audit, wdyt?

File: src/emailLogin/runMigration.js
Comment (kishore03109) [inline src/emailLogin/runMigration.js]: hey checking in here, this should be uncommented right?

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js]: Sure, that makes sense, moved it out in https://github.com/isomerpages/isomer-conversion-scripts/pull/56/commits/ad8f51847323111bd76b5136823e56138970caf8

File: src/emailLogin/runMigration.js
Comment (kishore03109) [inline src/emailLogin/runMigration.js:84]: Hey this should be `user.email.endsWith(domain))` rather than `user.email.endsWith(domain)).length > 0` right?

File: src/emailLogin/runMigration.js
Comment (kishore03109) [inline src/emailLogin/runMigration.js:80]: Hey I am quite confused here :(
While _all_ of our migration army users have whitelisted emails that end with gmail/ymail/hotmail accounts, not _all_ of those emails that end with gmail/ymail/hotmail are part of of our migration army correct? 

As part of the migration, shouldnt we treat migration army as power collaborators rather than admins, and potentially use github as a source of truth to map which repos they have collaborator access to? 

can huddle on this if typing is hard HAHA

File: src/emailLogin/runMigration.js
Comment (kishore03109) [inline src/emailLogin/runMigration.js:89]: may I get your help to output them to another file? 
Reason for this is because there are some redundant repos that have been created and no commits at all, was thinking we can take this opportunity to clean up our infra! 

what I am doing rn with Jackie is whatever repo that has no code, I am passing it to him to check if the folks still exist and if can I am deleting them. Perhaps we can do the same for repos with no CMS editors? 

PS. all repos have isomerAdmin as a collaborator, may I confirm that this if condition is indeed reachable?

File: src/emailLogin/runMigration.js
Comment (kishore03109) [inline src/emailLogin/runMigration.js:23]: Are we storing these logs in the cloud?

**Review by kishore03109 — CHANGES_REQUESTED** [review]

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js]: Yep, already fixed in 327e3dc

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js:84]: It's actually `whitelistedDomains.filter((domain) => user.email.endsWith(domain))` and then checking that at least one of the whitelist domains match (`.length > 0`)

Comment (kishore03109) [reply src/emailLogin/runMigration.js:84]: arh HAHA ok makes sense

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js:80]: I've checked with Jackson - the only whitelisted emails which end with generic emails (gmail, ymail, hotmail) are from our migration army, he's always ensured that agencies only send them specialised emails to whitelist.

I'm wondering about the migration army access actually - I decided to remove them so that agencies don't see them as collaborators later on after they're on the identity flow. We can chat more about this offline!

Comment (kishore03109) [reply src/emailLogin/runMigration.js:80]: oki catch you tomorrow sir Alex~

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js:89]: I'm actually filtering out all the isomer users already - so this is only to check which are the agency users to grant access to. What happens right now is

1. We retrieve the list of github ids mapped to the github team
2. We match the github ids to the appropriate isomer user (which means that agency users who have never logged in will have no matches here)
3. We add these people as site members

At the moment, I've been keeping track with the spreadsheet for any sites with this issue, but I'm thinking that this is a temporary measure - having no collaborators means that only our migration army has touched this repo so far, but I don't expect it to be this way closer to site launches, so I'm planning to rerun the script for these sites again on a later date.

I've also tried to keep track of any additional repos which have no code on the spreadsheet, though I've only found one so far - happy to chat more about this offline!

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js:23]: Nope, this is just for our own records in case we need to revert or something - I think we also might not want to push user data, so it'll be just to keep track of migrations locally

File: src/emailLogin/runMigration.js
Comment (kishore03109) [reply src/emailLogin/runMigration.js:89]: > having no collaborators means that only our migration army has touched this repo so far,

Hmm not necessarily right, might have some users that use one email for github and another for isomer? 

> I've also tried to keep track of any additional repos which have no code on the spreadsheet, though I've only 
found one so far - happy to chat more about this offline! 


eh lets combine this, ill add another column in the main spreadsheet

File: src/emailLogin/runMigration.js
Comment (kishore03109) [reply src/emailLogin/runMigration.js:23]: nono cloud like our private notion/our google doc hahah

Comment (alexanderleegs) [reply src/emailLogin/runMigration.js:23]: Will push to a private folder in our drive!

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js:80]: As discussed offline - we're going to only give agency users access via email login, since the majority of the sites will already be okay by the time we provide them with email login access. Any future changes that they want us to change which involves the migration army will involve us providing them with the email of the user to provide access, so that agencies can do their own user management in future

File: src/emailLogin/removeGithubAccess.js
Comment (alexanderleegs) [reply src/emailLogin/removeGithubAccess.js]: Ah yup didn't think about private repos, good catch! Have changed it to remove access entirely in https://github.com/isomerpages/isomer-conversion-scripts/pull/56/commits/eff5b6840753c7335c6f359f86ba9ab098ee9d32

File: src/emailLogin/runMigration.js
Comment (seaerchin) [reply src/emailLogin/runMigration.js:8]: i resolve this cos i trust you

File: src/emailLogin/removeGithubAccess.js
Comment (seaerchin) [inline src/emailLogin/removeGithubAccess.js]: #55 seems to have changed this to `GITHUB_ACCESS_TOKEN` fyi; also doesn't the token itself need to have sufficient perms to do the operations below? if so, a personal token might not cut it and i'd be in favour of calling it `GITHUB_ACCESS_TOKEN` or smth along those lines to convey the elevated perms required

File: src/emailLogin/removeGithubAccess.js
Comment (seaerchin) [inline src/emailLogin/removeGithubAccess.js]: should just beef up this msg with the `site`

File: src/emailLogin/runMigration.js
Comment (seaerchin) [inline src/emailLogin/runMigration.js]: shud log site

File: src/emailLogin/runMigration.js
Comment (seaerchin) [inline src/emailLogin/runMigration.js]: what happens if we get an error? will the connection be terminated?

**Review by seaerchin — COMMENTED** [review]
minor comments

File: src/emailLogin/runMigration.js
Comment (seaerchin) [reply src/emailLogin/runMigration.js:23]: actually if this is required, why not just commit to git rather than ignoring it?

File: src/emailLogin/removeGithubAccess.js
Comment (alexanderleegs) [reply src/emailLogin/removeGithubAccess.js]: Sure, changed in https://github.com/isomerpages/isomer-conversion-scripts/pull/56/commits/39c30003fe4a7c539d1264d016f22e0c0d6b7360

File: src/emailLogin/removeGithubAccess.js
Comment (alexanderleegs) [reply src/emailLogin/removeGithubAccess.js]: https://github.com/isomerpages/isomer-conversion-scripts/pull/56/commits/ca319a0cbd33cbca781829394f37b918de5af727

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js]: https://github.com/isomerpages/isomer-conversion-scripts/pull/56/commits/ca319a0cbd33cbca781829394f37b918de5af727

File: src/emailLogin/runMigration.js
Comment (alexanderleegs) [reply src/emailLogin/runMigration.js]: Not at the moment - i've changed it in https://github.com/isomerpages/isomer-conversion-scripts/pull/56/commits/3e715d8b4da6a37c00e0edaf40c60db9b1bc0425!

**Review by seaerchin — APPROVED** [review]
pls resolve @kishore03109's points before merging HAHA

**Review by kishore03109 — APPROVED** [review]

**Pull request merged**
